### PR TITLE
Add cluster-policy-controller container

### PR DESCRIPTION
**Pull request by sallyom**
- adds cluster-policy-controller container to static pod.

@deads2k 

related PRs
* rolebinding/role: https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/120
* lock  NamespaceSCCAllocationController: https://github.com/openshift/openshift-controller-manager/pull/28
* lock ClusterPolicyController: https://github.com/openshift/cluster-policy-controller/pull/3
* CI https://github.com/openshift/release/pull/5075

### CODE REVIEW

File: bindata/v4.1.0/kube-controller-manager/pod.yaml [hunk 1]
```diff
@@ -22,6 +22,23 @@ spec:
         sleep 1
       done
   containers:
+  - name: cluster-policy-controller
+    image: ${CLUSTER_POLICY_CONTROLLER_IMAGE}
```
Comment (sallyom) [inline bindata/v4.1.0/kube-controller-manager/pod.yaml]: :woman_shrugging:

File: bindata/v4.1.0/kube-controller-manager/pod.yaml [hunk 2]
```diff
@@ -22,6 +22,23 @@ spec:
...
+    - -v=2
+    resources:
+      requests:
+        memory: 100Mi
+        cpu: 100m
+    ports:
+      - containerPort: 8443
+    volumeMounts:
+      - mountPath: /etc/kubernetes/static-pod-resources
+        name: resource-dir
```
Comment (sallyom) [inline bindata/v4.1.0/kube-controller-manager/pod.yaml]: I need more info about this container, to get this correct

File: bindata/v4.1.0/kube-controller-manager/pod.yaml (hunk 1 above)
Comment (deads2k) [reply bindata/v4.1.0/kube-controller-manager/pod.yaml]: @soltysh you know how to plumb the imagestream from the payload?  If not, ask @sanchezl to help you get sorted out.

File: bindata/v4.1.0/kube-controller-manager/pod.yaml [hunk 3]
```diff
@@ -22,6 +22,23 @@ spec:
         sleep 1
       done
   containers:
+  - name: cluster-policy-controller
+    image: ${CLUSTER_POLICY_CONTROLLER_IMAGE}
+    imagePullPolicy: IfNotPresent
+    terminationMessagePolicy: FallbackToLogsOnError
+    command: ["cluster-policy-controller", "start"]
+    args:
+    - --config=/etc/kubernetes/static-pod-resources/configmaps/config/config.yaml
```
Comment (deads2k) [inline bindata/v4.1.0/kube-controller-manager/pod.yaml]: I wasn't expecting a config

File: bindata/v4.1.0/kube-controller-manager/pod.yaml [hunk 4]
```diff
@@ -22,6 +22,23 @@ spec:
         sleep 1
       done
   containers:
+  - name: cluster-policy-controller
+    image: ${CLUSTER_POLICY_CONTROLLER_IMAGE}
+    imagePullPolicy: IfNotPresent
+    terminationMessagePolicy: FallbackToLogsOnError
+    command: ["cluster-policy-controller", "start"]
+    args:
```
Comment (deads2k) [inline bindata/v4.1.0/kube-controller-manager/pod.yaml]: you need a `--kubeconfig` don't you?

File: bindata/v4.1.0/kube-controller-manager/pod.yaml (hunk 2 above)
Comment (deads2k) [reply bindata/v4.1.0/kube-controller-manager/pod.yaml]: > I need more info about this container, to get this correct

mount the same kubeconfig we use for the kube-controller-manager

File: bindata/v4.1.0/kube-controller-manager/pod.yaml [hunk 5]
```diff
@@ -22,6 +22,23 @@ spec:
...
+    - --config=/etc/kubernetes/static-pod-resources/configmaps/config/config.yaml
+    - -v=2
+    resources:
+      requests:
+        memory: 100Mi
+        cpu: 100m
+    ports:
```
Comment (deads2k) [inline bindata/v4.1.0/kube-controller-manager/pod.yaml]: you must have health checks before merge

File: bindata/v4.1.0/kube-controller-manager/pod.yaml [hunk 6]
```diff
@@ -22,6 +22,23 @@ spec:
...
+      - containerPort: 8443
```
Comment (deads2k) [inline bindata/v4.1.0/kube-controller-manager/pod.yaml]: this pod is hostnetwork.  You need to find something that doesn't conflict.  You may as well update the cluster-policy-controller to change the default.

File: manifests/0000_90_kube-controller-manager-operator_05_cluster-policy-controller-role.yaml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  namespace: openshift-kube-controller-manager
+  name: cluster-policy-controller-leader-lock
+rules:
+- apiGroups: [""]
+  resources: ["configmaps"]
+  resourceNames: ["cluster-policy-controller"]
```
Comment (deads2k) [inline manifests/0000_90_kube-controller-manager-operator_05_cluster-policy-controller-role.yaml]: this won't allow you to create.  You probably already have permissions, but if not, just grant yourself access to all the resource names.

File: manifests/0000_90_kube-controller-manager-operator_06_cluster-policy-controller-rolebinding.yaml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: cluster-policy-controller-leader-lock-cm
+  namespace: openshift-kube-controller-manager
+subjects:
+- kind: User
+  name: system:serviceaccount:openshift-controller-manager:openshift-controller-manager-sa
```
Comment (deads2k) [inline manifests/0000_90_kube-controller-manager-operator_06_cluster-policy-controller-rolebinding.yaml]: no.  This belongs in the openshift-controller-manager-operator eventually.  Wait until we're done here to open that.

File: manifests/0000_90_kube-controller-manager-operator_05_cluster-policy-controller-role.yaml (hunk 7 above)
Comment (sallyom) [reply manifests/0000_90_kube-controller-manager-operator_05_cluster-policy-controller-role.yaml]: oh, that's it! ok

File: manifests/0000_90_kube-controller-manager-operator_06_cluster-policy-controller-rolebinding.yaml (hunk 8 above)
Comment (sallyom) [reply manifests/0000_90_kube-controller-manager-operator_06_cluster-policy-controller-rolebinding.yaml]: ok

File: bindata/v4.1.0/kube-controller-manager/pod.yaml (hunk 2 above)
Comment (sallyom) [reply bindata/v4.1.0/kube-controller-manager/pod.yaml]: done

File: bindata/v4.1.0/kube-controller-manager/pod.yaml (hunk 3 above)
Comment (sallyom) [reply bindata/v4.1.0/kube-controller-manager/pod.yaml]: removed, added kubeconfig

File: bindata/v4.1.0/kube-controller-manager/pod.yaml (hunk 6 above)
Comment (sallyom) [reply bindata/v4.1.0/kube-controller-manager/pod.yaml]: updated, and looking for where that is

File: bindata/v4.1.0/kube-controller-manager/pod.yaml (hunk 5 above)
Comment (sallyom) [reply bindata/v4.1.0/kube-controller-manager/pod.yaml]: added

File: bindata/v4.1.0/kube-controller-manager/pod.yaml [hunk 9]
```diff
@@ -21,7 +21,50 @@ spec:
...
         sleep 1
       done
+  - name: wait-for-other-host-port
+    terminationMessagePolicy: FallbackToLogsOnError
+    image: ${CLUSTER_POLICY_CONTROLLER_IMAGE}
+    imagePullPolicy: IfNotPresent
+    command: ['/usr/bin/timeout', '30', "/bin/bash", "-c"]
+    args:
+    - |
+      echo -n "Waiting for port :10258 to be released."
```
Comment (deads2k) [inline bindata/v4.1.0/kube-controller-manager/pod.yaml]: this is the cloud controller manager port.  Choose something in a different range.

File: manifests/0000_90_kube-controller-manager-operator_05_cluster-policy-controller-role.yaml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  namespace: openshift-kube-controller-manager
+  name: cluster-policy-controller-leader-lock
```
Comment (deads2k) [inline manifests/0000_90_kube-controller-manager-operator_05_cluster-policy-controller-role.yaml]: if you don't bind this here, move it with the openshift-controller-manager

File: bindata/bootkube/bootstrap-manifests/kube-controller-manager-pod.yaml [hunk 11]
```diff
@@ -12,6 +12,33 @@ spec:
...
+  - name: cluster-policy-controller
+    image: quay.io/openshift/origin-cluster-policy-controller:v4.3
+    imagePullPolicy: {{ .ImagePullPolicy }}
+    ports:
+      - containerPort: 10357
+    command: ["cluster-policy-controller", "start"]
+    args:
+    - --config=/etc/kubernetes/config/{{ .ConfigFileName }}
+    - --kubeconfig=/etc/kubernetes/secrets/kubeconfig
+    - --alsologtostderr
```
Comment (sallyom) [inline bindata/bootkube/bootstrap-manifests/kube-controller-manager-pod.yaml]: @sanchezl for the command and args, can you confirm what's required here?
When I run `./cluster-policy-controller start`:
```
$ ./cluster-policy-controller start --help
Start the Cluster Policy Controller

Usage:
  cluster-policy-controller start [flags]

Flags:
      --config string   Location of the master configuration file to run from.
  -h, --help            help for start

Global Flags:
      --alsologtostderr                  log to standard error as well as files
---
```
does this controller need the `config` (and/or `kubeconfig`)? or is that leftover from openshift-controller-manager cmd.  thanks for your help.

File: bindata/v4.1.0/kube-controller-manager/pod.yaml (hunk 9 above)
Comment (sallyom) [reply bindata/v4.1.0/kube-controller-manager/pod.yaml]: fixed, it's 10357 now : )

File: manifests/0000_90_kube-controller-manager-operator_05_cluster-policy-controller-role.yaml (hunk 10 above)
Comment (sallyom) [reply manifests/0000_90_kube-controller-manager-operator_05_cluster-policy-controller-role.yaml]: fixed

File: manifests/0000_90_kube-controller-manager-operator_06_cluster-policy-controller-rolebinding.yaml (hunk 8 above)
Comment (sallyom) [reply manifests/0000_90_kube-controller-manager-operator_06_cluster-policy-controller-rolebinding.yaml]: done

File: bindata/bootkube/bootstrap-manifests/kube-controller-manager-pod.yaml (hunk 11 above)
Comment (sallyom) [reply bindata/bootkube/bootstrap-manifests/kube-controller-manager-pod.yaml]: looking at the pod the PR test cluster, I don't see this container in the static pod.. maybe it needs to be added here instead? 
https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/bindata/v4.1.0/kube-controller-manager/pod.yaml

**Comment by sallyom** [comment]
In the PR test-cluster openshift-kube-controller-manager static pod does not include the cluster-policy-controller container w/ these changes, hm..

File: bindata/bootkube/bootstrap-manifests/kube-controller-manager-pod.yaml (hunk 11 above)
Comment (sanchezl) [reply bindata/bootkube/bootstrap-manifests/kube-controller-manager-pod.yaml]: Looks like it still needs the config.
I created pull request to remove references to the part of the config no longer needed: https://github.com/openshift/cluster-policy-controller/pull/5.

Comment (sallyom) [reply bindata/bootkube/bootstrap-manifests/kube-controller-manager-pod.yaml]: cool thanks!

Comment (sanchezl) [reply bindata/bootkube/bootstrap-manifests/kube-controller-manager-pod.yaml]: I opened a PR to make the config optional (we were using an empty config file anyway).
https://github.com/openshift/cluster-policy-controller/pull/6

Comment (sallyom) [reply bindata/bootkube/bootstrap-manifests/kube-controller-manager-pod.yaml]: got it, thanks

**Comment by sallyom** [comment]
> In the PR test-cluster openshift-kube-controller-manager static pod does not include the cluster-policy-controller container w/ these changes, hm.

Looking at a test cluster, the static pod running in the cluster is defined with this: [bindata/v4.1.0-pod.yaml](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/bindata/v4.1.0/kube-controller-manager/pod.yaml) not [bindata/bootkube/bootstrap-maniftests pod.yaml](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/bindata/bootkube/bootstrap-manifests/kube-controller-manager-pod.yaml)

**Comment by sallyom** [comment]
Great, at least now my container definition is breaking the install.  Now we (@sanchezl @deads2k - i'll probably be bugging you ) can zero in on what's required. : )

**Comment by deads2k** [comment]
look in starter.go 

```
                        "message": "InstallerControllerDegraded: missing required resources: [configmaps: config-1,controller-manager-kubeconfig-1,kube-controller-cert-syncer-kubeconfig-1,kube-controller-manager-pod-1,service-ca-1,serviceaccount-ca-1, secrets: csr-signer-1,kube-controller-manager-client-cert-key-1,service-account-private-key-1]\nRevisionControllerDegraded: configmaps \"kube-controller-manager-pod\" not found\nStaticPodsDegraded: pods \"kube-controller-manager-ip-10-0-146-183.ec2.internal\" not found\nStaticPodsDegraded: pods \"kube-controller-manager-ip-10-0-130-118.ec2.internal\" not found\nStaticPodsDegraded: pods \"kube-controller-manager-ip-10-0-129-156.ec2.internal\" not found",

```

File: bindata/v4.1.0/kube-controller-manager/pod.yaml (hunk 3 above)
Comment (deads2k) [reply bindata/v4.1.0/kube-controller-manager/pod.yaml]: it already merged

File: bindata/v4.1.0/kube-controller-manager/pod.yaml [hunk 12]
```diff
@@ -21,7 +21,48 @@ spec:
...
+    args:
+    - |
+      echo -n "Waiting for port :10357 to be released."
+      while [ -n "$(lsof -ni :10357)" ]; do
+        echo -n "."
+        sleep 1
+      done
   containers:
+  - name: cluster-policy-controller-REVISION
+    image: registry.svc.ci.openshift.org/ocp/4.3:cluster-policy-controller
```
Comment (sallyom) [inline bindata/v4.1.0/kube-controller-manager/pod.yaml]: better/more correct way to reference `cluster-policy-controller` image? This unblocks my testing for now

File: bindata/v4.1.0/kube-controller-manager/pod.yaml (hunk 3 above)
Comment (sallyom) [reply bindata/v4.1.0/kube-controller-manager/pod.yaml]: yes, the container needs the config, w/out it I'm seeing (w/ local test, running the container manually in a test cluster) 
```
I1015 01:33:50.754554       1 policy_controller.go:41] Starting controllers on 0.0.0.0:8443 (v0.0.0-unknown)
F1015 01:33:50.754639       1 cmd.go:57] open /var/run/configmaps/client-ca/ca-bundle.crt: no such file or directory
```
I'm hoping somebody who has more experience with these controllers running in openshift-controller-manager pod can take a look, as I'm semi-unfamiliar w/ the actual controller code (although getting more familiar now)

**Comment by sallyom** [comment]
running a test with a cluster-policy-controller image built from this PR: https://github.com/openshift/cluster-policy-controller/pull/11

File: bindata/v4.1.0/kube-controller-manager/pod.yaml [hunk 13]
```diff
@@ -21,7 +21,54 @@ spec:
...
+    - |
+      echo -n "Waiting for port :10357 to be released."
+      while [ -n "$(lsof -ni :10357)" ]; do
+        echo -n "."
+        sleep 1
+      done
   containers:
+  - name: cluster-policy-controller-REVISION
+    #image: registry.svc.ci.openshift.org/ocp/4.3:cluster-policy-controller
+    image: quay.io/sallyom/cluster-policy-controller:new
```
Comment (sallyom) [inline bindata/v4.1.0/kube-controller-manager/pod.yaml]: DON'T MERGE WITH THIS, TESTING

Comment (sallyom) [reply bindata/v4.1.0/kube-controller-manager/pod.yaml]: removed

File: bindata/v4.1.0/kube-controller-manager/pod.yaml [hunk 14]
```diff
@@ -59,6 +71,38 @@ spec:
         path: healthz
       initialDelaySeconds: 10
       timeoutSeconds: 10
+  - name: cluster-policy-controller-REVISION
+    #image: registry.svc.ci.openshift.org/ocp/4.3:cluster-policy-controller
+    image: quay.io/sallyom/cluster-policy-controller:test
```
Comment (sallyom) [inline bindata/v4.1.0/kube-controller-manager/pod.yaml]: testing image

**Comment by sallyom** [comment]
closing, in favor of https://github.com/openshift/cluster-kube-controller-manager-operator/pull/297

**Pull request closed without merging**